src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -694,33 +694,65 @@ private void AssignPatternVariables(BoundPattern pattern)
         {
-            var result = base.VisitReturnStatementNoAdjust(node);
-
             Debug.Assert(!IsConditionalState);
-            if (node.ExpressionOpt != null && this.State.Reachable)
+
+            var expr = node.ExpressionOpt;
+            if (expr == null)
             {
-                TypeSymbolWithAnnotations returnType = this._currentMethodOrLambda?.ReturnType;
+                return null;
+            }
+
+            expr = RemoveImplicitConversions(expr);
+            var result = VisitRvalueWithResult(expr);
+
+            //if (this.State.Reachable) // PROTOTYPE(NullableReferenceTypes): Consider reachability?
+            if (expr.Type?.IsErrorType() == true)
+            {
+                return null;
+            }
+
+            TypeSymbolWithAnnotations returnType = GetReturnType(compilation, _currentMethodOrLambda);
+            if ((object)returnType != null)
+            {
+                TypeSymbolWithAnnotations unconvertedType = result.Type;
+                Conversion conversion = GenerateConversion(_conversions, expr, unconvertedType?.TypeSymbol, returnType.TypeSymbol);
+                TypeSymbolWithAnnotations resultType = InferResultNullability(expr, conversion, returnType.TypeSymbol, unconvertedType);
+
                 bool returnTypeIsNonNullable = IsNonNullable(returnType);
-                bool returnTypeIsUnconstrainedTypeParameter = IsUnconstrainedTypeParameter(returnType?.TypeSymbol);
+                bool returnTypeIsUnconstrainedTypeParameter = IsUnconstrainedTypeParameter(returnType.TypeSymbol);
                 bool reportedNullable = false;
                 if (returnTypeIsNonNullable || returnTypeIsUnconstrainedTypeParameter)
                 {
                     reportedNullable = ReportNullAsNonNullableReferenceIfNecessary(node.ExpressionOpt);
                 }
+
                 if (!reportedNullable)
                 {
-                    TypeSymbolWithAnnotations resultType = _result.Type;
                     if (IsNullable(resultType) && (returnTypeIsNonNullable || returnTypeIsUnconstrainedTypeParameter) ||
                         IsUnconstrainedTypeParameter(resultType?.TypeSymbol) && returnTypeIsNonNullable)
                     {
                         ReportStaticNullCheckingDiagnostics(ErrorCode.WRN_NullReferenceReturn, node.ExpressionOpt.Syntax);
                     }
+                    else if (!conversion.Exists)
+                    {
+                        ReportStaticNullCheckingDiagnostics(ErrorCode.WRN_NullabilityMismatchInAssignment, expr.Syntax, GetTypeAsDiagnosticArgument(unconvertedType?.TypeSymbol), returnType.TypeSymbol);
+                    }
                 }
-
-                ReportNullabilityMismatchInAssignmentIfNecessary(node.ExpressionOpt, node.ExpressionOpt.Type, returnType.TypeSymbol);
             }
 
-            return result;
+            return null;
+        }
```

Comment:
Not related to this PR: have we tested enumerator methods? There should be a similar situation since the type that you can `yield return` is not the return type of the method.

Reply:
#23701